ChattoAdditions/Source/Input/Photos/Photo/PhotosInputCellProvider.swift

```diff
@@ -0,0 +1,101 @@
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+*/
+
+import UIKit
+
+protocol PhotosInputCellProviderProtocol: class {
+    func cellForItem(at indexPath: IndexPath) -> UICollectionViewCell
+    func configureFullImageLoadingIndicator(at indexPath: IndexPath,
+                                            request: PhotosInputDataProviderImageRequestProtocol)
+}
+
+final class PhotosInputCellProvider: PhotosInputCellProviderProtocol {
+    private let reuseIdentifier = "PhotosCellProvider"
+    private let collectionView: UICollectionView
+    private let dataProvider: PhotosInputDataProviderProtocol
+    init(collectionView: UICollectionView, dataProvider: PhotosInputDataProviderProtocol) {
+        self.dataProvider = dataProvider
+        self.collectionView = collectionView
+        self.collectionView.register(PhotosInputCell.self, forCellWithReuseIdentifier: self.reuseIdentifier)
+    }
+
+    func cellForItem(at indexPath: IndexPath) -> UICollectionViewCell {
+        let cell = self.collectionView.dequeueReusableCell(withReuseIdentifier: self.reuseIdentifier, for: indexPath) as! PhotosInputCell
+        self.configureCell(cell, at: indexPath)
+        return cell
+    }
+
+    func configureFullImageLoadingIndicator(at indexPath: IndexPath,
+                                            request: PhotosInputDataProviderImageRequestProtocol) {
+        guard let cell = self.collectionView.cellForItem(at: indexPath) as? PhotosInputCell else { return }
+        self.configureCellForFullImageLoadingIfNeeded(cell, request: request)
+    }
+
+    private var previewRequests = [Int: PhotosInputDataProviderImageRequestProtocol]()
+    private var fullImageRequests = [Int: PhotosInputDataProviderImageRequestProtocol]()
+    private func configureCell(_ cell: PhotosInputCell, at indexPath: IndexPath) {
+        if let request = self.previewRequests[cell.hash] {
+            self.previewRequests[cell.hash] = nil
+            self.dataProvider.cancelImageRequest(request)
+        }
+        self.fullImageRequests[cell.hash] = nil
+        cell.hideProgressView()
```

Comment:
Can you do it in `prepareForReuse` function? Cell state should be clean before reuse.

Reply:
👌